### PR TITLE
Added filters to prevent the deletion from RUCSS

### LINKED ISSUE #4942: Allow disabling the automatic clearing of used CSS
**Is your feature request related to a problem? Please describe.**

A customer requested a way to prevent the automatic clearing of used CSS given that the site doesn't change often.

**Describe the solution you'd like**

We can have a filter that will prevent used CSS from being automatically cleared.

We should preserve the ability to clear the used CSS globally or on a per-page basis using the admin bar menu item.

**Describe alternatives you've considered**

We can use a helper plugin like the following to do this already:
https://drive.google.com/uc?id=1JDPFtXUBzYfFWlgDbkBxA64xKriWqZD6

The downside is that we would need to update the helper plugin in the future if we decide to add/remove triggers for the automatic removal of used CSS.

**Additional context**

**Ticket:** https://secure.helpscout.net/conversation/1849389806/337707/

### PR DESCRIPTION
**Pull request by CrochetFeve0251**
## Description
Add two filters allowing control of the deletion from the RUCSS.

Fixes #4942

## Type of change

Please delete options that are not relevant.

- [x] Enhancement (non-breaking change which improves an existing functionality)

## Is the solution different from the one proposed during the grooming?

No.

## How Has This Been Tested?

- [x] Test locally

# Checklist:

Please delete the options that are not relevant.

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: inc/Engine/Optimization/RUCSS/Admin/Database.php [hunk 1]
```diff
@@ -62,22 +62,26 @@ public function truncate_used_css_table() : bool {
...
+	 *
 	 * @return void
 	 */
-	public function delete_old_used_css() {
+	public function delete_old_used_css( $delete_delay = 1 ) {
 		if ( $this->rucss_usedcss_table->exists() ) {
-			$this->rucss_usedcss_table->delete_old_used_css();
+			$this->rucss_usedcss_table->delete_old_used_css( $delete_delay );
 		}
 	}
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Admin/Database.php]: It's better to use the filter value directly in the method itself

File: inc/Engine/Optimization/RUCSS/Admin/Database.php [hunk 2]
```diff
@@ -62,22 +62,26 @@ public function truncate_used_css_table() : bool {
...
+	 *
 	 * @return void
 	 */
-	public function delete_old_resources() {
+	public function delete_old_resources( $delete_delay = 1 ) {
 		if ( $this->rucss_resources_table->exists() ) {
-			$this->rucss_resources_table->delete_old_items();
+			$this->rucss_resources_table->delete_old_items( $delete_delay );
 		}
 	}
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Admin/Database.php]: No need to b e updated as it's no longer used

Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Admin/Database.php]: No need to b e updated as it's no longer used

File: inc/Engine/Optimization/RUCSS/Admin/Subscriber.php [hunk 3]
```diff
@@ -126,8 +127,14 @@ public function cron_clean_rows() {
 			return;
 		}
 
-		$this->database->delete_old_used_css();
-		$this->database->delete_old_resources();
+		$delete_delay = apply_filters( 'rocket_rucss_css_delete_delay', 1 );
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Admin/Subscriber.php]: use the `is_rucss_deletion_activated()` here too instead

Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Admin/Subscriber.php]: use the `is_rucss_deletion_activated()` here too instead

File: inc/Engine/Optimization/RUCSS/Admin/Subscriber.php [hunk 4]
```diff
@@ -254,6 +261,12 @@ public function delete_used_css_on_update_or_delete( $post_id ) {
 			return;
 		}
 
+		$deletion_activated = $this->is_rucss_deletion_activated();
+
+		if ( ! $deletion_activated ) {
+			return;
+		}
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Admin/Subscriber.php]:
```suggestion
		$deletion_activated = $this->is_rucss_deletion_activated();

		if ( ! $deletion_activated ) {
			return;
		}
```
```suggestion
		if ( ! $this->is_rucss_deletion_activated() ) {
			return;
		}
```

Same for the other cases, it's not necessary to assign to a variable here

Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Admin/Subscriber.php]:
```suggestion
		$deletion_activated = $this->is_rucss_deletion_activated();

		if ( ! $deletion_activated ) {
			return;
		}
```
```suggestion
		if ( ! $this->is_rucss_deletion_activated() ) {
			return;
		}
```

Same for the other cases, it's not necessary to assign to a variable here

File: inc/Engine/Optimization/RUCSS/Admin/Subscriber.php [hunk 5]
```diff
@@ -770,4 +795,32 @@ public function cancel_queues() {
...
+	}
+
+	/**
+	 * Returns default value from Delete Delay.
+	 *
+	 * @return int
+	 */
+	public function set_delete_delay() {
+		return 1;
+	}
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Admin/Subscriber.php]: All of this is unnecessary, we already set the default values in the filter declaration

Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Admin/Subscriber.php]: All of this is unnecessary, we already set the default values in the filter declaration

File: inc/Engine/Optimization/RUCSS/Admin/Subscriber.php [hunk 6]
```diff
@@ -770,4 +795,32 @@ public function cancel_queues() {
...
 		delete_transient( 'rocket_rucss_as_tables_count' );
 	}
+
+	/**
+	 * Call the filter to check if the deletion from RUCSS is activated.
+	 *
+	 * @return mixed|null
+	 */
+	protected function is_rucss_deletion_activated() {
+		return apply_filters( 'rocket_rucss_deletion_activated', true );
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Admin/Subscriber.php]: Add docblock for the filter https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#4-hooks-actions-and-filters

And force a `bool` return

Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Admin/Subscriber.php]: Add docblock for the filter https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#4-hooks-actions-and-filters

File: inc/Engine/Optimization/RUCSS/Admin/Subscriber.php [hunk 7]
```diff
@@ -770,4 +795,32 @@ public function cancel_queues() {
 	public function delete_as_tables_transient_on_tools_page() {
 		delete_transient( 'rocket_rucss_as_tables_count' );
 	}
+
+	/**
+	 * Call the filter to check if the deletion from RUCSS is activated.
+	 *
+	 * @return mixed|null
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Admin/Subscriber.php]:
```suggestion
	 * @return bool
```

Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Admin/Subscriber.php]:
```suggestion
	 * @return bool
```

File: inc/Engine/Optimization/RUCSS/Database/Tables/Resources.php [hunk 8]
```diff
@@ -65,9 +65,11 @@ protected function set_schema() {
 	/**
 	 * Delete all resources which were not accessed in the last month.
 	 *
+	 * @param int $delete_delay number of months before deleting.
+	 *
 	 * @return bool|int
 	 */
-	public function delete_old_items() {
+	public function delete_old_items( $delete_delay = 1 ) {
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Database/Tables/Resources.php]: No need to update this method, it's not used anymore

Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Database/Tables/Resources.php]: No need to update this method, it's not used anymore

File: inc/Engine/Optimization/RUCSS/Database/Tables/UsedCSS.php [hunk 9]
```diff
@@ -69,9 +69,11 @@ protected function set_schema() {
 	/**
 	 * Delete all used_css which were not accessed in the last month.
 	 *
+	 * @param int $delete_delay number of months before deleting.
+	 *
 	 * @return bool|int
 	 */
-	public function delete_old_used_css() {
+	public function delete_old_used_css( $delete_delay = 1 ) {
```
Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Database/Tables/UsedCSS.php]: Use the filter value here, and make sure it's a positive int and not zero value before using it in the query

Comment (remyperona) [inline inc/Engine/Optimization/RUCSS/Database/Tables/UsedCSS.php]: Use the filter value here, and make sure it's a positive int and not zero value before using it in the query

**Review by remyperona — CHANGES_REQUESTED** [review]

File: inc/Engine/Optimization/RUCSS/Admin/Subscriber.php (hunk 3 above)
Comment (CrochetFeve0251) [reply inc/Engine/Optimization/RUCSS/Admin/Subscriber.php:130]: @Tabrisrp the idea was to let the user control the period of time `is_rucss_deletion_activated` returns a boolean.

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@CrochetFeve0251 Thanks for the PR.
Clear used CSS of this URL from page preview is not working in the following cases:
1- Old page whose used CSS > 1 month while the related filter `rocket_rucss_delete_interval` is set to 0
2- When the filter `rocket_rucss_deletion_enabled` is set to false then we try clear used CSS of this post
Can you please check?

**Comment by Mai-Saad** [comment]
@CrochetFeve0251 Clear used CSS of this URL is working fine now. But if `rocket_rucss_deletion_enabled` is set to false, the edit/quick edit post will clear the used CSS of that post. Can you please check?

**Review by mostafa-hisham — APPROVED** [review]

**Review by Mai-Saad — APPROVED** [review]
Working as expected, failed test due to issue on the trunk.
Note:
Using invalid values for filter i.e Null, "" ,...etc,  the filter will prevent deletion of used CSS (not the default behavior which allows deletion)   => fine as is as per discussion with @piotrbak 
[testrail-report-411.pdf](https://github.com/wp-media/wp-rocket/files/9535024/testrail-report-411.pdf)

**Pull request merged**